File: src/matlab/spacedyn_v2r0/aw.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% The SpaceDyn, a MATLAB toolbox for Space and Mobile Robots.
% (C)1998 The Space Robotics Lab. directed by Kazuya Yoshida,
% Tohoku University, Japan.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%    version 1.0 // Oct.4, 1999, Last modification by K.Yoshida
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%	Aw( w0 ) returns a 3x3 transformation representing a 
%	rotation about the vector w0.
%
%   2002.2.27 H.Hamano modified to version up
%
%	See also CX, CY, CZ.

function E0 = aw( SV )

global d_time


if ( norm(SV.w0)==0 )
   E0 = eye(3);
   
else
   th = norm(SV.w0) * d_time;
   w = SV.w0 ./ norm(SV.w0);
   
   E0 =[ cos(th)+w(1)^2*(1-cos(th)) ...
         w(1)*w(2)*(1-cos(th))-w(3)*sin(th) ...
         w(3)*w(1)*(1-cos(th))+w(2)*sin(th);
         
         w(1)*w(2)*(1-cos(th))+w(3)*sin(th) ...
         cos(th)+w(2)^2*(1-cos(th)) ...
         w(3)*w(2)*(1-cos(th))-w(1)*sin(th);
         
         w(3)*w(1)*(1-cos(th))-w(2)*sin(th) ...
         w(3)*w(2)*(1-cos(th))+w(1)*sin(th) ...
         cos(th)+w(3)^2*(1-cos(th)) ];
   
end

%%%EO

File: aw.py
import numpy as np
from math import cos
from math import sin
from Get_global_value import d_time


# aw( w0 ) returns a 3x3 transformation representing a rotation about the vector w0.
def aw(omega0):
    omega0_norm = np.linalg.norm(omega0)
    transform_about_omega0 = np.zeros((3, 3))
    if omega0_norm == 0:
        transform_about_omega0 = np.eye(3)
    else:
        theta = omega0_norm * d_time
        unit_omega0 = np.true_divide(omega0, omega0_norm)
        transform_about_omega0[0, 0] = cos(theta) + unit_omega0[0] ** 2 * (1-cos(theta))
        transform_about_omega0[0, 1] = unit_omega0[0] * unit_omega0[1] * (1-cos(theta)) - unit_omega0[2] * sin(theta)
        transform_about_omega0[0, 2] = unit_omega0[2] * unit_omega0[0] * (1-cos(theta)) + unit_omega0[1] * sin(theta)
        transform_about_omega0[1, 0] = unit_omega0[0] * unit_omega0[1] * (1-cos(theta)) + unit_omega0[2] * sin(theta)
        transform_about_omega0[1, 1] = cos(theta) + unit_omega0[1] ** 2 * (1-cos(theta))
        transform_about_omega0[1, 2] = unit_omega0[2] * unit_omega0[1] * (1-cos(theta)) - unit_omega0[0] * sin(theta)
        transform_about_omega0[2, 0] = unit_omega0[2] * unit_omega0[0] * (1-cos(theta)) - unit_omega0[1] * sin(theta)
        transform_about_omega0[2, 1] = unit_omega0[2] * unit_omega0[1] * (1-cos(theta)) + unit_omega0[0] * sin(theta)
        transform_about_omega0[2, 2] = cos(theta) + unit_omega0[2] ** 2 * (1-cos(theta))

    return transform_about_omega0









